server/app/routes/oauth2/authenticate_api.rb

```diff
@@ -0,0 +1,71 @@
+require 'uri'
+require_relative '../../services/auth_provider'
+require_relative '../../helpers/token_authentication_helper'
+
+module OAuth2Api
+
+  # The /authenticate endpoint creates a AuthorizationRequest object which
+  # holds the "state" parameter that is passed along the way during the
+  # browser auth choreography. It also holds the original redirect-uri which
+  # normally points to the CLI's localhost webserver.
+  #
+  # Unless the auth provider config has been set, it will respond with HTTP 501.
+  #
+  # If the redirect uri is valid, the user will be redirected to auth
+  # provider's authorize url.
+  #
+  # You can also pass in invite_code that you will get when creating invites.
+  class AuthenticateApi < Roda
+
+    include OAuth2Api::Common
+    include TokenAuthenticationHelper
+
+    route do |r|
+      r.get do
+
+        params = request.params
+
+        unless AuthProvider.valid?
+          mime_halt(
+            501,
+            'server_error',
+            'Authentication provider not configured, use: kontena master auth-provider config'
+          )
+          return
+        end
+
+        unless params['redirect_uri']
+          mime_halt(400, 'invalid_request', 'Missing redirect_uri')
+          return
+        end
+
+        redirect_uri = URI.parse(params['redirect_uri'])
+
+        # Only allow redirect to client's localhost or back to same server's /code'
+        unless (redirect_uri.host && redirect_uri.host == 'localhost') || (redirect_uri.host.nil? && redirect_uri.path == '/code')
+          mime_halt(400, 'invalid_request', 'Invalid redirect_uri') and return
+        end
+
+        invite_code = params['invite_code']
+        if invite_code.kind_of?(String) && invite_code.length > 4
```

Comment:
The code is ignored. The callback will try to find the user by using email. But it should probably halt if you try an invalid code. Fixed now.